### PR TITLE
Add networking diagnostic app

### PR DESCRIPTION
**Pull request by miz060**
# Description

Add the networking diagnostic app.

## Type of change

- New content

# Checklist:

- [x] My content has a file named README.md, which is based on the appropriate readme [template](https://github.com/Azure/azure-sphere-gallery/tree/main/Templates)
- [x] I have performed a self-review of my own contribution
- [x] I have provided comments where appropriate
- [x] I have updated the repository's [README.md](https://github.com/Azure/azure-sphere-gallery/blob/main/README.md) to index this project
- [x] I have added/updated license(s) for this project as appropriate
- [x] I have permission to publish this content (in case the content was collaborative)
- [x] I have verified that my content does not contain sensitive information (PII, Microsoft PI, etc.) and is safe to open source

### CODE REVIEW

**Review by tokenstar — CHANGES_REQUESTED** [review]
thanks for the PR! I'll make some comments inline and start a naming conversation offline

**Review by tokenstar — CHANGES_REQUESTED** [review]

File: DiagnosticApp/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    if (err == -1) {
+        Log_Debug("Error setting interface state %d\n",errno);
+        return -1;
+    }
+   ```
+
+## Running the code
+
+### Build and run the app
+To build and run this app, follow the instructions in [Build a high-level application](https://docs.microsoft.com/en-us/azure-sphere/install/qs-blink-application?tabs=windows%2Ccliv2beta&pivots=vs-code).
```
[the remark this answers is not in the document]
Comment (miz060) [reply DiagnosticApp/README.md]: got it. updated.

File: DiagnosticApp/README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Networking Diagnostic Application
+
+**Note:** In Progress. Currently supporting DNS Resolver and Custom NTP Test.
```
[the remark this answers is not in the document]
Comment (miz060) [reply DiagnosticApp/README.md]: I agree that the suggested paragraph is better. Updated.

File: DiagnosticApp/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# Networking Diagnostic Application
```
[the remark this answers is not in the document]
Comment (miz060) [reply DiagnosticApp/README.md]: Project name updated as OSNetworkRequirementChecker-HLApp.

File: DiagnosticApp/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        Custom NTP Time Server Sync list:
+        Index: 0, Name: ###.###.###.###, UTC time before sync: t1, after sync: t2
+
+      ......
+
+## Next steps
+
+### Project expectations
+This app is an ongoing project. More features will added in the future.
```
[the remark this answers is not in the document]
Comment (miz060) [reply DiagnosticApp/README.md]: I see. Updated as suggested.

**Review by tokenstar — APPROVED** [review]

File: OSNetworkRequirementChecker-HLApp/src/dns-helper.c [hunk 5]
```diff
@@ -0,0 +1,519 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cawhitworth) [inline OSNetworkRequirementChecker-HLApp/src/dns-helper.c:1]: This code appears to be based off an older version of the DNSServiceDiscovery sample from the azure-sphere-samples repository - it contains a bug where misleading results will be shown if there is more than one message is receieved. Please see the changes in [this PR](https://github.com/Azure/azure-sphere-samples/pull/196/files#diff-d889dd24f89bdfb65d958c52bf2b71d51b9055f9c707889b2d7573877b38df94) for a fix.

**Review by cawhitworth — CHANGES_REQUESTED** [review]
Hi - this looks fine, but please port over the fix to the DNS parser code from the DNSServiceDiscovery sample before completing.

File: OSNetworkRequirementChecker-HLApp/src/dns-helper.c (hunk 5 above)
Comment (miz060) [reply OSNetworkRequirementChecker-HLApp/src/dns-helper.c:1]: I see. thanks for the pointer! updated.

File: OSNetworkRequirementChecker-HLApp/src/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License.
+
+cmake_minimum_required(VERSION 3.10)
+
+project(DiagnosticApp C)
```
Comment (ongchris) [inline OSNetworkRequirementChecker-HLApp/src/CMakeLists.txt:6]: Should this name be updated too?

File: OSNetworkRequirementChecker-HLApp/src/app_manifest.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
+  "SchemaVersion": 1,
+  "Name": "DiagnosticApp",
```
Comment (ongchris) [inline OSNetworkRequirementChecker-HLApp/src/app_manifest.json:3]: Should this name be updated too?

**Review by ongchris — APPROVED** [review]

File: OSNetworkRequirementChecker-HLApp/src/dns-helper.c [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+            break;
+        }
+        case ns_t_srv: {
+            // Parse the SRV record and populate the port and host fields in instance details as per
+            // DNS SRV record specification: https://tools.ietf.org/rfc/rfc2782.txt
+            // SRV record format: Priority|  Weight |   Port  |     Target
+            //                   (2 Bytes)|(2 Bytes)|(2 Bytes)|(Remaining Bytes)
+            const char *data = ns_rr_rdata(rr);
+            (*instanceDetails)->port =
+                (uint16_t)ns_get16(data + sizeof(uint16_t) + sizeof(uint16_t));
```
Comment (cawhitworth) [inline OSNetworkRequirementChecker-HLApp/src/dns-helper.c:149]: This needs removing - it should only be set if the port/host are unset.

**Review by cawhitworth — APPROVED** [review]
Approving - please fix the remaining issue before completion.

File: OSNetworkRequirementChecker-HLApp/src/app_manifest.json [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  "ComponentId": "8128f290-f4b1-4986-945c-1868d38e33a5",
+  "EntryPoint": "/bin/app",
+  "CmdArgs": [],
+  "Capabilities": {
+    "AllowedConnections": [ 
+      "eastus-prod-azuresphere.azure-devices.net",
```
Comment (ongchris) [inline OSNetworkRequirementChecker-HLApp/src/app_manifest.json:9]: Can you add these endpoints from https://docs.microsoft.com/en-us/azure-sphere/network/ports-protocols-domains?
global.azure-devices-provisioning.net
global.azure-devices-provisioning.net

Also need to add these new ones?
            "anse.azurewatson.microsoft.com",
            "sphereblobeus.azurewatson.microsoft.com",
            "sphereblobweus.azurewatson.microsoft.com"

File: OSNetworkRequirementChecker-HLApp/src/CMakeLists.txt (hunk 6 above)
Comment (miz060) [reply OSNetworkRequirementChecker-HLApp/src/CMakeLists.txt]: Yes, updated now.

File: OSNetworkRequirementChecker-HLApp/src/app_manifest.json (hunk 7 above)
Comment (miz060) [reply OSNetworkRequirementChecker-HLApp/src/app_manifest.json]: Yes, updated.

File: OSNetworkRequirementChecker-HLApp/src/dns-helper.c (hunk 8 above)
Comment (miz060) [reply OSNetworkRequirementChecker-HLApp/src/dns-helper.c]: oh yes I missed that. Fixed.

File: OSNetworkRequirementChecker-HLApp/src/app_manifest.json (hunk 9 above)
Comment (miz060) [reply OSNetworkRequirementChecker-HLApp/src/app_manifest.json:10]: updated.

**Pull request merged**